src/cc/BPF.h

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2016 Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <linux/bpf.h>
+#include <memory>
+#include <sstream>
+#include <string>
+
+#include "bcc_syms.h"
+#include "bpf_module.h"
+#include "BPFTable.h"
+#include "libbpf.h"
+
+namespace ebpf {
+
+class BPF {
+ public:
+  static const int BPF_MAX_STACK_DEPTH = 127;
+
+  explicit BPF(
+    const std::string& bpf_program,
+    unsigned int debug_flag = 0,
+    const char* cflags[] = nullptr,
+    int ncflags = 0
+  ) :
+    bpf_module_(std::make_shared<BPFModule>(debug_flag))
+  {
+    if (bpf_module_->load_string(bpf_program, cflags, ncflags) != 0)
+      throw std::runtime_error("Unable to initialize BPF program");
+  };
+
+  ~BPF();
+
+  void attach_kprobe(
+    const std::string& kernel_func,
+    const std::string& probe_func,
+    bool ret_probe = false,
+    pid_t pid = -1, int cpu = 0, int group_fd = -1,
+    perf_reader_cb cb = nullptr, void* cb_cookie = nullptr
```

Comment:
We talked about this right? You said that was some legacy behavior (return 1 for perf event listener). Should we just not keep it and always construct empty reader?